### PR TITLE
[Frontend] Dynamic RoPE scaling

### LINKED ISSUE #4334: [Feature]: Set RoPE scaling parameters dynamically
### 🚀 The feature, motivation and pitch

As it was implemented in #555, specifying RoPE parameters is only available from the model's `config.json`, and I haven't found a way to set it dynamically in my code. Is there currently a way of doing this?

Related to #910.

### Alternatives

Right now, unless providing a modified `config.json` (which is very inconvenient in my setup), I haven't found an alternative. I've tried monkey patching `vllm.transformers_utils.config.get_config` function to no avail (Ray uses it in a way I don't understand).

### Additional context

For context - specific to my setup - I'm using 2 GPUs to run a quantized Llama-3-70B (casperhansen's). Thus vLLM is using Ray.

### PR DESCRIPTION
**Pull request by sasha0552**
In #555, @WoosukKwon removed dynamic specifying of RoPE scaling with the comment:
> As we discussed offline, I removed rope_scaling from ModelConfig and EngineArgs. Now rope_scaling is always read from the model's config.json.

I don't understand why this feature was removed, so this PR brings it back. Specifying the RoPE scaling on the command line is very useful, because otherwise we have to manually modify the `config.json` that can be managed by huggingface - so each model has to be forked to properly set a different RoPE scaling.

FIX #4334

<details>
<summary>RoPE scaling allows to use higher context without further fine-tuning</summary>

Summarization using `meta-llama/Meta-Llama-3-8B-Instruct` (has a native context of 8192 tokens) with `type` = `linear` and `factor` = `2.0`:

```text
ChatCompletion(id='cmpl-11ec2bc71a734a1eba13eaca5e0c54d9', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content="The article provides information about CUDA, a parallel computing platform and programming model created by NVIDIA. It allows software developers to use graphics processing units (GPUs) for general-purpose computing, rather than just graphics processing graphics. CUDA is a proprietary, but has been adopted by many developers. The article covers CUDA's history, features, technical specifications, capabilities, and usage.\n\nCUDA's features and capabilities include:\n\n* Parallel programming model: CUDA allows for parallel processing of tasks on multiple threads and blocks\n* Memory management: CUDA's memory hierarchy includes registers, shared memory, and global memory\n* Execution: CUDA code can be executed on multiple GPUs and CPUs\n* Interoperability: CUDA supports various programming languages, including C, C++, Fortran, and Python\n* APIs: CUDA has a low-level API (CUDA Driver) and high-level (CUDA) API, with libraries and runtime\n\nCUDA has several advantages, including:\n\n* Higher performance, especially for computationally intensive tasks\n* Larger memory bandwidth and storage\n* Better power efficiency\n\nCUDA's compute capabilities include:\n\n* Wide range of compute capabilities (1.0 to 11.1)\n* Different memory types (shared, global, texture, and constant)\n* Instructions (e.g., ALU, INT, FP, and FP16)\n* Number of AL lanes, texture mapping units, and scheduling\n* Warp size and block sizes\n\nCUDA is used for various applications, including:\n\n* Accelerated rendering, video, encryption, and decryption\nBioinformatics, medical simulations, machine learning\nNeural network, proteins, cryptography, and more\n\nCUDA competes with other GPU computing stacks, such as Intel's OneAPI and AMD's ROCm.", role='assistant', function_call=None, tool_calls=None), stop_reason=128009)], created=1715044970, model='meta-llama/Meta-Llama-3-8B-Instruct', object='chat.completion', system_fingerprint=None, usage=CompletionUsage(completion_tokens=344, prompt_tokens=12648, total_tokens=12992))
```
</details>

### CODE REVIEW

File: vllm/engine/arg_utils.py [hunk 1]
```diff
@@ -320,6 +322,10 @@ def add_cli_args(
                             'None, we assume the model weights are not '
                             'quantized and use `dtype` to determine the data '
                             'type of the weights.')
+        parser.add_argument('--rope-scaling',
+                            default=None,
+                            type=json.loads,
+                            help='RoPE scaling configuration')
```
Comment (mgoin) [inline vllm/engine/arg_utils.py:328]: Can you share an example of what the CLI format is like for this arg in the help function?

File: vllm/transformers_utils/config.py [hunk 2]
```diff
@@ -41,6 +42,8 @@ def get_config(model: str,
         config = config_class.from_pretrained(model,
                                               revision=revision,
                                               code_revision=code_revision)
+    if rope_scaling is not None:
+        config.update({"rope_scaling": rope_scaling})
```
Comment (mgoin) [inline vllm/transformers_utils/config.py:46]: Worth putting an info log here to saying "Updating rope_scaling from xxx to yyy"

**Comment by tom-doerr** [comment]
Could this get merged? Especially with Llama 3 being very tolerant of RoPE scaling this is very useful

**Comment by mgoin** [comment]
Sure this would be great to get in. Currently we don't have a test for it though and it might be too intensive with a real model, so I would like to see at least a unit test implemented. @sasha0552 could you add a test to [`tests/test_config.py`](https://github.com/vllm-project/vllm/blob/main/tests/test_config.py)?

**Comment by sasha0552** [comment]
@mgoin test added. Can you review? The failures are not related to this PR.

**Review by mgoin — APPROVED** [review]
Thank you @sasha0552! This is great. We just merged a fix for the failing tests, so please rebase and the tests should pass https://github.com/vllm-project/vllm/pull/4944

**Comment by sasha0552** [comment]
@mgoin can you merge? All tests passed.

**Pull request merged**